src/charm.py

```diff
@@ -2,163 +2,279 @@
+                self.log.error(traceback.format_exc())
+                raise ErrorWithStatus("Failed to replan", BlockedStatus)
+
+    def _update_profiles_container(self, event) -> None:
+        """Updates the Profiles Pebble configuration layer if changed."""
+        if not self.profiles_container.can_connect():
+            self.unit.status = WaitingStatus("Waiting to connect to Profiles container")
+            event.defer()
+            return
+
+        try:
+            self._check_leader()
+            self.unit.status = MaintenanceStatus("Configuring Profiles layer")
+            self._update_profiles_layer()
+        except ErrorWithStatus as error:
+            raise error
+        self.unit.status = MaintenanceStatus("Profiles layer configured")
+
+    def _update_kfam_container(self, event) -> None:
+        """Updates the kfam Pebble configuration layer if changed."""
+        if not self.profiles_container.can_connect():
+            self.unit.status = WaitingStatus("Waiting to connect to kfam container")
+            event.defer()
+            return
+
+        try:
+            self._check_leader()
+            self.unit.status = MaintenanceStatus("Configuring kfam layer")
+            update_layer(
+                self._kfam_container_name,
+                self.kfam_container,
+                self._kfam_pebble_layer,
+                self.log,
+            )
+        except ErrorWithStatus as error:
+            raise error
+        self.unit.status = MaintenanceStatus("kfam layer configured")
+
+    def _on_kubeflow_profiles_ready(self, event):
+        """Define and start a workload using the Pebble API.
+        Learn more about Pebble layers at https://github.com/canonical/pebble
+        """
+        try:
+            self._update_profiles_container(event)
+
+        except ErrorWithStatus as e:
+            self.model.unit.status = e.status
+            if isinstance(e.status, BlockedStatus):
+                self.log.error(str(e.msg))
+            else:
+                self.log.info(str(e.msg))
+
+    def _on_kfam_ready(self, event):
+        """Define and start a workload using the Pebble API.
+        Learn more about Pebble layers at https://github.com/canonical/pebble
+        """
+        try:
+            self._update_kfam_container(event)
+
+        except ErrorWithStatus as e:
+            self.model.unit.status = e.status
+            if isinstance(e.status, BlockedStatus):
+                self.log.error(str(e.msg))
+            else:
+                self.log.info(str(e.msg))
+
+    def _on_remove(self, event):
+        self.unit.status = MaintenanceStatus("Removing k8s resources")
+        manifests = self.k8s_resource_handler.render_manifests()
+        try:
+            delete_many(self.k8s_resource_handler.lightkube_client, manifests)
+        except ApiError as e:
+            self.log.warning(f"Failed to delete resources: {manifests} with: {e}")
+            raise e
```

Comment:
Is raising here the desired action?  If we raise here, the charm will go into an error state.  Is the preference here that we hit an error state (and therefore the application is not removed), or should it *try* to remove these items but if it fails it logs the issues but proceeds anyway?

There's no right or wrong, just making sure it is doing the one you prefer.  Sometimes, especially in a remove hook, we want to *try* to clean up what we can but we're ok with failing and continuing on

Reply:
that depends on the consequences of not cleaning up the resources, would that affect another deploy per say?
also I think we should reevaluate this same behavior in other charms, dashboard and seldon that I know of.

Reply:
Yeah this feels like a good discussion in a tech talk, and then us adopting some general guidelines.  I think we've gotten stung by errors in `remove` hooks before, because they block `juju` from removing all the other stuff that it should (eg: juju wont delete all the rest of the things it should until the remove hook is resolved), which iirc caused bigger problems.  But I might be remembering this wrong

Reply:
added to tech topics

Reply:
After discussions with Juju and internally, it was decided that the best practice is to:
* attempt to delete things, catching any errors
* if we fail to delete for a known and acceptable reason (for example, object no longer exists), then warn in logs but do not set Error state
* if we fail for some odd reason (some unexpected error), that might warrant setting Error

The main reason for this is that setting a charm to an error state in a Remove hook is a hard thing for users to navigate out of, and often is worse than the errors we catch